core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java

```diff
@@ -576,13 +587,118 @@ private static boolean representAsInternalType(Type type) {
    * Handle decimal type specifically with explicit type conversion
    */
   private static Expression convertAssignableType(
-      Expression argument,  Type targetType) {
+      Expression argument, Type targetType) {
     if (targetType != BigDecimal.class) {
       return argument;
     }
     return convert(argument, targetType);
   }
 
+  public static MethodCallExpression call(Class clazz, String methodName,
+      List<? extends Expression> arguments) {
+    return call(clazz, methodName, arguments, null);
+  }
+
+  /**
+   * The powerful version of {@code org.apache.calcite.linq4j.tree.Expressions#call(
+   * Type, String, Iterable<? extends Expression>)}. Try best effort to convert the
+   * accepted arguments to match parameter type.
+   *
+   * @param clazz Class against which method is invoked
+   * @param methodName Name of method
+   * @param arguments argument expressions
+   * @param targetExpression target expression
+   *
+   * @return MethodCallExpression that call the given name method
+   * @throws RuntimeException if no suitable method found
+   */
+  public static MethodCallExpression call(Class clazz, String methodName,
+       List<? extends Expression> arguments, Expression targetExpression) {
+    Class[] argumentTypes = Types.toClassArray(arguments);
+    try {
+      Method candidate = clazz.getMethod(methodName, argumentTypes);
+      return Expressions.call(targetExpression, candidate, arguments);
+    } catch (NoSuchMethodException e) {
+      for (Method method : clazz.getMethods()) {
+        if (method.getName().equals(methodName)) {
+          final boolean varArgs = method.isVarArgs();
+          final Class[] parameterTypes = method.getParameterTypes();
+          if (Types.allAssignable(varArgs, parameterTypes, argumentTypes)) {
+            return Expressions.call(targetExpression, method, arguments);
+          }
+          // fall through
+          final List<? extends Expression> typeMatchedArguments =
+              matchMethodParameterTypes(varArgs, parameterTypes, arguments);
+          if (typeMatchedArguments != null) {
+            return Expressions.call(targetExpression, method, typeMatchedArguments);
+          }
+        }
+      }
+      throw new RuntimeException("while resolving method '" + methodName
+          + Arrays.toString(argumentTypes) + "' in class " + clazz, e);
+    }
+  }
+
+  private static List<? extends Expression> matchMethodParameterTypes(boolean varArgs,
+      Class[] parameterTypes, List<? extends Expression> arguments) {
+    if (varArgs) {
```

Comment:
minor: maybe we could simplify this piece of code by unifying into a single 'if'?
```
if ( (varArgs  && arguments.size() < parameterTypes.length - 1) ||
     (!varArgs && arguments.size() != parameterTypes.length) ) {
  return null;
}
```

Reply:
Thanks, I agree, addressed.